### PR TITLE
Service Refactor, Upload Avatars

### PR DESCRIPTION
**Pull request by MatkoTom**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant
motivation and context. List any dependencies that are required for this change.

Branch contains work for uploading avatars via our service class. Also updated the service class to be more flexible and to make future event handling easier.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can
reproduce. Please also list any relevant details for your test configuration. All testing was done
manually by following, plainly, functionality requirements. If there are some special edge cases or
special ways that things need to be tested they will be mentioned below.

Dev testing

**Test Configuration**:

* Firmware version:
* Hardware: SAMSUNG Galaxy S22
* SDK: Android 14

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat_details/ChatDetailsFragment.kt [hunk 1]
```diff
@@ -632,20 +583,27 @@ class ChatDetailsFragment : BaseFragment() {
...
-                    null,
-                    false,
-                    null,
-                    roomWithUsers.room.roomId,
-                    null,
-                    null
-                )
+
+            avatarData = FileData(
+                currentPhotoLocation,
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat_details/ChatDetailsFragment.kt:588]: can you maybe add: fileUri = currentPhotoLocation, etc.?

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat_details/ChatDetailsFragment.kt [hunk 2]
```diff
@@ -632,20 +583,27 @@ class ChatDetailsFragment : BaseFragment() {
...
+                false,
+                null,
+                roomWithUsers.room.roomId,
+                null,
+                null
             )
+
+            if (bound) {
+                CoroutineScope(Dispatchers.Default).launch {
+                    fileUploadService.uploadAvatar(avatarData!!, isGroup = true)
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat_details/ChatDetailsFragment.kt:602]: can we remove this "!!" ?

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/settings/SettingsFragment.kt [hunk 3]
```diff
@@ -286,70 +289,37 @@ class SettingsFragment : BaseFragment() {
...
-                        override fun fileCanceledListener(messageId: String?) {
-                            // Ignore
-                        }
-
-                    })
-            }
+
+            avatarData = FileData(
+                currentPhotoLocation,
+                Const.JsonFields.AVATAR_TYPE,
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/settings/SettingsFragment.kt:303]: same thing here

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/settings/SettingsFragment.kt [hunk 4]
```diff
@@ -429,8 +399,53 @@ class SettingsFragment : BaseFragment() {
...
+
+            override fun avatarUploadFinished() {
+                requireActivity().runOnUiThread {
+                    binding.profilePicture.flProgressScreen.visibility = View.GONE
+                }
+            }
+        })
+
+        CoroutineScope(Dispatchers.Default).launch {
+            fileUploadService.uploadAvatar(avatarData!!, isGroup = false)
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/settings/SettingsFragment.kt:444]: can we remove this "!!"?

File: app/src/main/java/com/clover/studio/spikamessenger/utils/helpers/UploadService.kt [hunk 5]
```diff
@@ -94,7 +86,53 @@ class UploadService : Service() {
...
+                    }
+                    callbackListener?.avatarUploadFinished()
+                    stopForeground(STOP_FOREGROUND_REMOVE)
+                    stopSelf()
+                    return
+                }
+
+                override fun fileCanceledListener(messageId: String?) {
+                    // avatar upload cannot be cancelled
+                }
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/utils/helpers/UploadService.kt:131]: why?

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt [hunk 6]
```diff
@@ -1612,4 +1547,94 @@ class ChatMessagesFragment : BaseFragment() {
...
+                    uriPairList.clear()
+                    uploadedFiles.clear()
+                    unsentMessages.clear()
+                    return
+                }
+            }
+
+            override fun uploadError(description: String) {
+                // TODO Something went wrong, maybe implement an error notification or dialog
+            }
```
[the remark this answers is not in the document]
Comment (ivolarevic) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt:1628]: Also, if we dont use this method in ChatMessagesFragment can we remove it?

**Review by ivolarevic — CHANGES_REQUESTED** [review]
Please check comments and commit fixes.

File: app/src/main/java/com/clover/studio/spikamessenger/utils/helpers/UploadService.kt (hunk 5 above)
Comment (MatkoTom) [reply app/src/main/java/com/clover/studio/spikamessenger/utils/helpers/UploadService.kt:131]: It is currently using an indeterminate spinner instead of a progress one. And there is no cancel option implemented there.

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (MatkoTom) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt:1628]: Removed

**Review by ivolarevic — APPROVED** [review]
LGTM

**Review by ivolarevic — APPROVED** [review]
LGTM
